File: android/app/src/main/java/app/unimusic/MusicKitAuthorizationPlugin.java
package app.unimusic;

import androidx.activity.result.ActivityResult;

import com.apple.android.sdk.authentication.AuthenticationFactory;
import com.apple.android.sdk.authentication.AuthenticationManager;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "MusicKitAuthorization")
public class MusicKitAuthorizationPlugin extends Plugin {
    private AuthenticationManager authManager;

    @Override
    public void load() {
        // This has to be initialized here instead of in field declaration
        // because otherwise Capacitor doesn't load the plugin correctly
        authManager = AuthenticationFactory.createAuthenticationManager(
                getContext()
        );
        super.load();
    }

    private String getDeveloperToken() {
        return getContext().getString(R.string.developer_token);
    }

    @PluginMethod()
    public void authorize(PluginCall call) {
        var intent = authManager.createIntentBuilder(getDeveloperToken())
                .setHideStartScreen(true)
                .build();

        startActivityForResult(call, intent, "finishAuthorization");
    }

    @ActivityCallback()
    private void finishAuthorization(PluginCall call, ActivityResult resultIntent) {
        if (resultIntent == null) {
            call.reject("MusicKit authorization failed");
            return;
        }

        var result = authManager.handleTokenResult(resultIntent.getData());
        if (result.isError()) {
            call.reject("MusicKit authorization failed:" + result.getError());
            return;
        }

        JSObject tokens = new JSObject();
        tokens.put("developerToken", getDeveloperToken());
        tokens.put("musicUserToken", result.getMusicUserToken());
        call.resolve(tokens);
    }
}

File: android/app/src/main/java/app/unimusic/MusicKitAuthorizationPlugin.kt
package app.unimusic

import androidx.activity.result.ActivityResult
import com.apple.android.sdk.authentication.AuthenticationFactory
import com.apple.android.sdk.authentication.AuthenticationManager
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "MusicKitAuthorization")
class MusicKitAuthorizationPlugin : Plugin() {
    private var authManager: AuthenticationManager? = null

    override fun load() {
        // This has to be initialized here instead of in field declaration
        // because otherwise Capacitor doesn't load the plugin correctly
        authManager = AuthenticationFactory.createAuthenticationManager(context)
        super.load()
    }

    private val developerToken: String
        get() = context.getString(R.string.developer_token)

    @PluginMethod
    fun authorize(call: PluginCall) {
        val intent = authManager!!.createIntentBuilder(developerToken)
            .setHideStartScreen(true)
            .build()

        startActivityForResult(call, intent, "finishAuthorization")
    }

    @ActivityCallback
    private fun finishAuthorization(call: PluginCall, resultIntent: ActivityResult?) {
        if (resultIntent == null) {
            call.reject("MusicKit authorization failed")
            return
        }

        val result = authManager!!.handleTokenResult(resultIntent.data)
        if (result.isError) {
            call.reject("MusicKit authorization failed:" + result.getError())
            return
        }

        val tokens = JSObject()
        tokens.put("developerToken", developerToken)
        tokens.put("musicUserToken", result.musicUserToken)
        call.resolve(tokens)
    }
}
